### PR TITLE
React Part 4 - initial draft version for feedback

### LINKED ISSUE #1337: React Part 4 - Apps
_Child task of Epic #1301_

The previous three lessons covered props, state, and component hierarchies. This will build on that to create a TODO list app with an in-memory data store, meaning all data will be lost on full page refresh. But we want to reinforce the existing content before getting into async/ajax.

Use our [Lesson Template v2](https://github.com/Techtonica/curriculum/blob/master/_templates/lesson-v2.md)

## Topics to Cover
- How to use `create-react-app` to create a skeleton app
  - A brief note as to why it's so complex / slow but that it's nice with live-reload etc
- Guided practice should be a TODO list app
- Once you have created the entry form and a way to mark items complete, add additional features as independent practice. This is the perfect time to differentiate the mastered data vs. a temporary array used for the active display
  - filtering completed out
  - different sorts, A-Z, Z-A, `creationDate` ASC or DESC
  - display the creationDate nicely using `moment`, which can cover how to add packages
- Independent practice can also include deploying to Heroku with the CRA build-pack

This lesson should go in the `./react-js/react-part-4-apps.md` and a link added to the main README.md along with the other React lessons.

### PR DESCRIPTION
**Pull request by vegetabill**
Closes #1337 

Feedback wanted!

Tried to keep this mostly limited to CRA and leave best practices etc to the final lesson, Part 5, see React epic: #1301

### CODE REVIEW

File: react-js/react-part4-full-apps.md [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+- You can organize your React components and CSS any way you want and consistently use `import` to refer to them (no more confusion about `require` vs `import`)
+- Quickly add npm packages using `yarn add`
+- Easily deploy to the cloud. You can use the `build` command to package all your code up
+- Lots of help available on Stack Overflow. Because many users, especially beginner engineers, are using CRA, they have probably encountered the same problem as you and are posting about it!
+
+### Objectives
+
+The important areas you should leave this lesson knowing.
+
+[ ] How does `create-react-app` compare to what we've been using so far?
```
Comment (alodahl) [inline react-js/react-part4-full-apps.md:27]:
```suggestion
- [ ] How does `create-react-app` compare to what we've been using so far?
```

Comment (alodahl) [reply react-js/react-part4-full-apps.md:27]: you'll want these for a checkbox

File: react-js/react-part4-full-apps.md [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+- CRA is only for toy apps and _real engineers_ setup React from scratch
+  - Actually, many professional projects leverage CRA or another starting point template like [Next.js](https://nextjs.org/) when beginning applications. Even if you want to customize the app in some ways, the general setup is usually a good starting point for many types of projects and **it will not make your portfolio project seem amateur to use it**
+- You **must** use CRA to build React apps
+  - Not true, but it is strongly recommended you build experience using it first before trying to do it yourself without CRA. Although it's much easier, especially for a beginner, many companies have ofted to create a customized setup so they have full control over all aspects of the application
+
+## Guided Practice
+
+Let's setup a small app.
+
+- `npx create-react-app todo-cra`
```
Comment (alodahl) [inline react-js/react-part4-full-apps.md:56]:
```suggestion
- `[npx](https://bambielli.com/til/2018-10-06-npx/) create-react-app todo-cra`
```

File: react-js/react-part4-full-apps.md [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+- _grab a coffee while it installs everything_
+- `cd todo-cra`
+- `yarn start`
+
+The app should open in a new browser tab automatically and show a spinning atom logo.
+
+- `code .` to open the generated project in your editor
+- Find `App.js` in the `src` folder
+- Change the contents of `<p>Hello Techtonica!</p>` and save
```
Comment (alodahl) [inline react-js/react-part4-full-apps.md:65]:
```suggestion
- Change the contents to: `<p>Hello Techtonica!</p>` and save
```

File: react-js/react-part4-full-apps.md [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+class Todo extends React.Component {
+  render() {
+    const { todo } = this.props;
+    return <div>{todo.text}</div>;
+  }
+}
+
+class App extends React.Component {
+  state = {
```
Comment (alodahl) [inline react-js/react-part4-full-apps.md:121]: I'm afraid state isn't this simple...

Comment (alodahl) [reply react-js/react-part4-full-apps.md:121]: https://reactjs.org/docs/react-component.html#constructor

File: react-js/react-part4-full-apps.md [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+};
+```
+
+Now let's update the `<Todo/>` component to render `isCompleted` as a checkbox. For attributes see [MDN Checkbox](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox).
+
+```jsx
+class Todo extends React.Component {
+  render() {
+    const {
+      todo: { text, isCompleted }
```
Comment (alodahl) [inline react-js/react-part4-full-apps.md:162]: Do they know enough about destructuring for this?

File: react-js/react-part4-full-apps.md [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+#### Style the Todo Items
+
+If they are complete, style them to indicate that, perhaps by graying them out or using strikethrough.
+
+#### Add Input Form
+
+To create new todos, add an input form that will create a new todo item and add it to the state so it shows up in the list.
+
+#### Add `reactstrap` npm package to your CRA project
```
Comment (alodahl) [inline react-js/react-part4-full-apps.md:327]: Do you think React Styling should come before or after this outline?  Does it matter?  https://github.com/Techtonica/curriculum/blob/main/react-js/styling-react.md

**Review by alodahl — CHANGES_REQUESTED** [review]
This is so great, thank you! 👏🏼

Some questions:

File: react-js/react-part4-full-apps.md (hunk 5 above)
Comment (vegetabill) [reply react-js/react-part4-full-apps.md]: Ooo fair. Looks like maybe no. Simplifying to not be more than one level deep

https://github.com/Techtonica/curriculum/blob/main/javascript/destructuring.md

File: react-js/react-part4-full-apps.md (hunk 4 above)
Comment (vegetabill) [reply react-js/react-part4-full-apps.md]: Oops good catch! I ran the code but might've "improved" it after pasting it here.

File: react-js/react-part4-full-apps.md (hunk 6 above)
Comment (vegetabill) [reply react-js/react-part4-full-apps.md:329]: Not sure where we planned for it to go in the schedule.... but I can at least link it up. I'd say a dedicated styling lesson can come later, maybe before Eventonica React.

#1160 added it during GSSOC.

Comment (vegetabill) [reply react-js/react-part4-full-apps.md:329]: While we're on it, those example code snippets looks a bit weird, so I opened an issue to polish that lesson up: #1503

**Pull request merged**